.github/workflows/go.yml

```diff
@@ -369,3 +369,41 @@ jobs:
       - name: Test
         shell: bash
         run: ci/scripts/go_test.sh $(pwd)
+
+  linux-arm:
+    name: ARM64 Debian 11 GO ${{ matrix.go }}
+    runs-on: ["arm", "linux"]
+    if: ${{ !contains(github.event.pull_request.title, 'WIP') }}
+    timeout-minutes: 60
+    strategy:
+      matrix:
+        go: [1.17]
+    env:
+      GO: ${{ matrix.go }}
+      ARROW_ENABLE_TIMING_TESTS: "OFF"
+      ARCH: arm64v8
+      ARROW_CI_MODULES: "GO"
+      DOCKER_IMAGE_ID: debian-go
+    steps:
+      - name: Checkout Arrow
+        uses: actions/checkout@v3
+        with:
+          fetch-depth: 0
+          submodules: recursive
+      - name: Setup Python
+        run: |
+          sudo apt install -y --no-install-recommends python3 python3-pip
```

Comment:
Can we unify this job to the existing `docker` job by using `apt` in the existing `docker` job too?

Reply:
you mean something like?
```
-      - name: Setup Python
-        run: |
-          sudo apt install -y --no-install-recommends python3 python3-pip
-      - name: Setup Archery
-        run: python3 -m pip install -e dev/archery[docker]
       - name: Execute Docker Build
         env:
           ARCHERY_DOCKER_USER: ${{ secrets.DOCKERHUB_USER }}
           ARCHERY_DOCKER_PASSWORD: ${{ secrets.DOCKERHUB_TOKEN }}
-        run: archery docker run debian-go
+        run: |
+          sudo apt install -y --no-install-recommends python3 python3-pip
+          python3 -m pip install -e dev/archery[docker]
+          archery docker run debian-go
```

Reply:
I didn't try it but something like the following:

```diff
diff --git a/.github/workflows/go.yml b/.github/workflows/go.yml
index 59c23c2e7b..02d750cd12 100644
--- a/.github/workflows/go.yml
+++ b/.github/workflows/go.yml
@@ -42,20 +42,30 @@ permissions:
 jobs:
 
   docker:
-    name: AMD64 Debian 11 Go ${{ matrix.go }}
-    runs-on: ubuntu-latest
+    name: ${{ matrix.arch-label }} Debian 11 Go ${{ matrix.go }}
+    runs-on: ${{ matrix.runs-on }}
     if: ${{ !contains(github.event.pull_request.title, 'WIP') }}
     timeout-minutes: 60
     strategy:
       fail-fast: false
       matrix:
-        go: [1.17, 1.18]
         include:
-          - go: 1.17
+          - arch-label: AMD64
+            arch: amd64
+            go: 1.17
+            runs-on: ubuntu-latest
             staticcheck: v0.2.2
-          - go: 1.18
+          - arch-label: AMD64
+            arch: amd64
+            go: 1.18
+            runs-on: ubuntu-latest
             staticcheck: v0.3.3
+          - arch-label: ARM64
+            arch: arm64v8
+            go: 1.17
+            runs-on: ["self-hosted", "arm", "linux"]
     env:
+      ARCH: ${{ matrix.arch }}
       GO: ${{ matrix.go }}
       STATICCHECK: ${{ matrix.staticcheck }}
     steps:
@@ -65,41 +75,51 @@ jobs:
           fetch-depth: 0
           submodules: recursive
       - name: Setup Python
-        uses: actions/setup-python@v4
-        with:
-          python-version: '3.10'
+        run: |
+          sudo apt install -y --no-install-recommends python3 python3-pip
       - name: Setup Archery
-        run: pip install -e dev/archery[docker]
+        run: python3 -m pip install -e dev/archery[docker]
       - name: Execute Docker Build
         env:
           ARCHERY_DOCKER_USER: ${{ secrets.DOCKERHUB_USER }}
           ARCHERY_DOCKER_PASSWORD: ${{ secrets.DOCKERHUB_TOKEN }}
         run: archery docker run debian-go
       - name: Docker Push
-        if: success() && github.event_name == 'push' && github.repository == 'apache/arrow'
+        if: >-
+          success() &&
+          github.event_name == 'push' &&
+          github.repository == 'apache/arrow'
         env:
           ARCHERY_DOCKER_USER: ${{ secrets.DOCKERHUB_USER }}
           ARCHERY_DOCKER_PASSWORD: ${{ secrets.DOCKERHUB_TOKEN }}
         continue-on-error: true
         run: archery docker push debian-go
       - name: Install Go ${{ matrix.go }} for Benchmarks
-        if: success() && github.event_name == 'push' && github.repository == 'apache/arrow'
+        if: >-
+          success() &&
+          matrix.arch == 'amd64' &&
+          github.event_name == 'push' &&
+          github.repository == 'apache/arrow'
         uses: actions/setup-go@v3
         with:
           go-version: ${{ matrix.go }}
           cache: true
           cache-dependency-path: go/go.sum
       - name: Run Benchmarks
-        if: success() && github.event_name == 'push' && github.repository == 'apache/arrow'
+        if: >-
+          success() &&
+          matrix.arch == 'amd64' &&
+          github.event_name == 'push' &&
+          github.repository == 'apache/arrow'
         env:
           CONBENCH_URL: https://conbench.ursa.dev
           CONBENCH_EMAIL: ${{ secrets.CONBENCH_EMAIL }}
           CONBENCH_PASSWORD: ${{ secrets.CONBENCH_PASS }}
           CONBENCH_REF: ${{ github.ref_name }}
-          CONBENCH_MACHINE_INFO_NAME: amd64-debian-11
+          CONBENCH_MACHINE_INFO_NAME: ${{ matrix.arch }}-debian-11
         run: |
-          pip install benchadapt@git+https://github.com/conbench/conbench.git@main#subdirectory=benchadapt/python
-          python ci/scripts/go_bench_adapt.py
+          python3 -m pip install benchadapt@git+https://github.com/conbench/conbench.git@main#subdirectory=benchadapt/python
+          python3 ci/scripts/go_bench_adapt.py
 
   docker_cgo:
     name: AMD64 Debian 11 GO ${{ matrix.go }} - CGO
@@ -373,41 +393,3 @@ jobs:
       - name: Test
         shell: bash
         run: ci/scripts/go_test.sh $(pwd)
-
-  linux-arm:
-    name: ARM64 Debian 11 GO ${{ matrix.go }}
-    runs-on: ["self-hosted", "linux"]
-    if: ${{ !contains(github.event.pull_request.title, 'WIP') }}
-    timeout-minutes: 60
-    strategy:
-      matrix:
-        go: [1.17]
-    env:
-      GO: ${{ matrix.go }}
-      ARROW_ENABLE_TIMING_TESTS: "OFF"
-      ARCH: arm64v8
-      ARROW_CI_MODULES: "GO"
-      DOCKER_IMAGE_ID: debian-go
-    steps:
-      - name: Checkout Arrow
-        uses: actions/checkout@v3
-        with:
-          fetch-depth: 0
-          submodules: recursive
-      - name: Setup Python
-        run: |
-          sudo apt install -y --no-install-recommends python3 python3-pip
-      - name: Setup Archery
-        run: python3 -m pip install -e dev/archery[docker]
-      - name: Execute Docker Build
-        env:
-          ARCHERY_DOCKER_USER: ${{ secrets.DOCKERHUB_USER }}
-          ARCHERY_DOCKER_PASSWORD: ${{ secrets.DOCKERHUB_TOKEN }}
-        run: archery docker run debian-go
-      - name: Docker Push
-        if: success() && github.event_name == 'push' && github.repository == 'apache/arrow'
-        env:
-          ARCHERY_DOCKER_USER: ${{ secrets.DOCKERHUB_USER }}
-          ARCHERY_DOCKER_PASSWORD: ${{ secrets.DOCKERHUB_TOKEN }}
-        continue-on-error: true
-        run: archery docker push debian-go

```

Reply:
Thanks @kou ! Sorry I missed what you meant on your initial comment. Thank you very much for the diff! I've pushed the changes to this PR: https://github.com/apache/arrow/pull/34512